### PR TITLE
Fixed LUT application | Added new stuff

### PR DESCRIPTION
**Pull request by AgnesMartinez**
Changed the way LUT was applied to pixel data

Added Image Brightness control using ImageEnhance from PIL

New create_GIF method to save a videoclip from DICOM files with multiple frames.

### CODE REVIEW

**Comment by darcymason** [comment]
This sounds interesting - I've been rather busy lately, but will try to have a look in the next few days, if no one else does...

**Comment by AgnesMartinez** [comment]
PIL's ImageEnhance module includes Color,Contrast and Sharpness too.

File: viewers/pydicom_PIL.py [hunk 1]
```diff
@@ -25,36 +23,44 @@
...
+    from PIL import Image,ImageEnhance
     have_PIL = True
 except ImportError:
     have_PIL = False
 
+#apply_modality_lut() function comes with pydicom module
 try:
-    import numpy as np
-    have_numpy = True
+    from pydicom.pixel_data_handlers.util import apply_modality_lut
```
Comment (mrbean-bremen) [inline viewers/pydicom_PIL.py:35]: I think it is sufficient to just import this without the try/except - after all, this is about `pydicom` contributions, so it is expected that `pydicom` is installed.

File: viewers/pydicom_PIL.py [hunk 2]
```diff
@@ -25,36 +23,44 @@
...
 
-def get_LUT_value(data, window, level):
+def get_LUT_value(dataset):
     """Apply the RGB Look-Up Table for the given
-       data and window/level value."""
-    if not have_numpy:
-        raise ImportError("Numpy is not available."
-                          "See http://numpy.scipy.org/"
+       dataset"""
+    if not have_lut:
```
Comment (mrbean-bremen) [inline viewers/pydicom_PIL.py:44]: see above - not needed

File: viewers/pydicom_PIL.py [hunk 3]
```diff
@@ -63,45 +69,149 @@ def get_PIL_image(dataset):
...
 
+    try:
+        #if pixel data has only one frame
+        im = Image.fromarray(image).convert(mode)
+
+    except:
+        #When pixel data has multiple frames, output the first one
+        im = Image.fromarray(image[0]).convert(mode)
+
+    if not brightness_factor == 1.0:
```
Comment (mrbean-bremen) [inline viewers/pydicom_PIL.py:107]: maybe `if brightness_factor != 1.0` :)

File: viewers/pydicom_PIL.py [hunk 4]
```diff
@@ -63,45 +69,149 @@ def get_PIL_image(dataset):
...
+        frame = Image.fromarray(item).convert(mode)
+
+        enhancer = ImageEnhance.Brightness(frame)
+
+        better_frame = enhancer.enhance(brightness_factor)
+
+        data.append(better_frame)
+    
+    #default brightness
+    gif.save(f'./{filename}.gif', save_all=True, quality=100 , append_images=data, optimize=True, duration=15, loop=0)
```
Comment (mrbean-bremen) [inline viewers/pydicom_PIL.py:217]: Looks like a nice addition to me.

**Review by mrbean-bremen — COMMENTED** [review]
For me this looks good, but I would wait for @darcymason to have another look.

File: viewers/pydicom_PIL.py (hunk 1 above)
Comment (AgnesMartinez) [reply viewers/pydicom_PIL.py:35]: Agree but i didnt want to refactor too much

File: viewers/pydicom_PIL.py [hunk 5]
```diff
@@ -25,36 +23,44 @@
...
-                        [0, 255, lambda data: ((data - (level - 0.5)) /
-                         (window - 1) + 0.5) * (255 - 0)])
+    #Pixel Data
+    pixel_data = dataset.pixel_array
+
+    return apply_modality_lut(pixel_data,dataset)
+
 
+def get_PIL_image(dataset,brightness_factor = 1.0):
+    """Get Image object from Python Imaging Library(PIL)
```
Comment (darcymason) [inline viewers/pydicom_PIL.py:56]: Just style, but should be space after comma and no spaces around the = for a function argument if not using typing notation

File: viewers/pydicom_PIL.py (hunk 3 above)
Comment (darcymason) [reply viewers/pydicom_PIL.py:107]: Agree with @mrbean-bremen mrbean

File: viewers/pydicom_PIL.py [hunk 6]
```diff
@@ -63,45 +69,149 @@ def get_PIL_image(dataset):
...
+
+        return better_im
+
+    
+    #return default brightness
     return im
 
 
 def show_PIL(dataset):
     """Display an image using the Python Imaging Library (PIL)"""
```
Comment (darcymason) [inline viewers/pydicom_PIL.py:122]: Shouldn't this also have a brightness factor to pass through to `get_PIL_image`?

File: viewers/pydicom_PIL.py [hunk 7]
```diff
@@ -63,45 +69,149 @@ def get_PIL_image(dataset):
...
     im = get_PIL_image(dataset)
+
     im.show()
+
+
+def create_GIF(dataset,brightness_factor = 1.0,filename="clip"):
+    """Create a GIF using the Python Imaging Library (PIL)
```
Comment (darcymason) [inline viewers/pydicom_PIL.py:129]: Style: spacing issues again

File: viewers/pydicom_PIL.py [hunk 8]
```diff
@@ -63,45 +69,149 @@ def get_PIL_image(dataset):
...
+            frame = Image.fromarray(item).convert(mode)
+
+            enhancer = ImageEnhance.Brightness(frame)
+
+            better_frame = enhancer.enhance(brightness_factor)
+
+            data.append(better_frame)
+
+        better_gif.save(f'./{filename}.gif', save_all=True, quality=100 , append_images=data, optimize=True, duration=15, loop=0)
+
```
Comment (darcymason) [inline viewers/pydicom_PIL.py:201]: I would suggest not to have the `./` here so user can pass a full file path into the function.  Also would be good to allow `Path` objects, i.e. use `str(filename)` here which would provide the proper pathname if is a `Path`.

Similarly with other save line below.

And ... perhaps can check if the `.gif` is already in the name and only add if necessary.

**Review by darcymason — COMMENTED** [review]
Just some style and minor code suggestions...

**Comment by darcymason** [comment]
Archiving contrib repository, closing issues first...

**Pull request closed without merging**